File: wear/src/main/java/dev/ungaaaabungaaa/relay/security/RelayCloudCrypto.kt
package dev.ungaaaabungaaa.relay.security

import java.nio.ByteBuffer
import java.nio.charset.StandardCharsets
import java.security.PrivateKey
import java.security.PublicKey
import java.util.Base64
import javax.crypto.Cipher
import javax.crypto.KeyAgreement
import javax.crypto.Mac
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec

data class RelayRoutingFields(
    val version: Int,
    val messageId: String,
    val accountId: String,
    val hostId: String,
    val senderId: String,
    val recipientId: String,
    val sentAt: Long,
    val sequence: Long,
)

data class RelayTunnelEnvelope(
    val version: Int,
    val messageId: String,
    val accountId: String,
    val hostId: String,
    val senderId: String,
    val recipientId: String,
    val sentAt: Long,
    val sequence: Long,
    val nonce: String,
    val ciphertext: String,
) {
    fun routing() = RelayRoutingFields(
        version,
        messageId,
        accountId,
        hostId,
        senderId,
        recipientId,
        sentAt,
        sequence,
    )
}

object RelayCloudCrypto {
    private const val INFO = "relay-e2ee-v1"

    fun deriveRootKey(
        privateKey: PrivateKey,
        peerPublicKey: PublicKey,
        pairingSessionNonce: ByteArray,
    ): ByteArray {
        require(pairingSessionNonce.size >= 16)
        val agreement = KeyAgreement.getInstance("ECDH")
        agreement.init(privateKey)
        agreement.doPhase(peerPublicKey, true)
        val sharedSecret = agreement.generateSecret()
        val pseudorandomKey = hmac(pairingSessionNonce, sharedSecret)
        return hmac(
            pseudorandomKey,
            INFO.toByteArray(StandardCharsets.UTF_8) + byteArrayOf(1),
        ).copyOf(32)
    }

    fun encrypt(
        routing: RelayRoutingFields,
        plaintext: ByteArray,
        rootKey: ByteArray,
        nonce: ByteArray,
    ): RelayTunnelEnvelope {
        require(rootKey.size == 32)
        require(nonce.size == 12)
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(
            Cipher.ENCRYPT_MODE,
            SecretKeySpec(rootKey, "AES"),
            GCMParameterSpec(128, nonce),
        )
        cipher.updateAAD(canonicalAAD(routing))
        return RelayTunnelEnvelope(
            routing.version,
            routing.messageId,
            routing.accountId,
            routing.hostId,
            routing.senderId,
            routing.recipientId,
            routing.sentAt,
            routing.sequence,
            Base64.getUrlEncoder().withoutPadding().encodeToString(nonce),
            Base64.getUrlEncoder().withoutPadding()
                .encodeToString(cipher.doFinal(plaintext)),
        )
    }

    fun decrypt(envelope: RelayTunnelEnvelope, rootKey: ByteArray): ByteArray {
        try {
            val cipher = Cipher.getInstance("AES/GCM/NoPadding")
            cipher.init(
                Cipher.DECRYPT_MODE,
                SecretKeySpec(rootKey, "AES"),
                GCMParameterSpec(
                    128,
                    Base64.getUrlDecoder().decode(envelope.nonce),
                ),
            )
            cipher.updateAAD(canonicalAAD(envelope.routing()))
            return cipher.doFinal(
                Base64.getUrlDecoder().decode(envelope.ciphertext),
            )
        } catch (_: Exception) {
            throw SecurityException("Relay envelope authentication failed")
        }
    }

    fun canonicalAAD(routing: RelayRoutingFields): ByteArray {
        require(routing.version == 1)
        require(routing.sequence > 0)
        val json = buildString {
            append("""{"version":${routing.version},"messageId":""")
            append(jsonString(routing.messageId))
            append(""","accountId":""")
            append(jsonString(routing.accountId))
            append(""","hostId":""")
            append(jsonString(routing.hostId))
            append(""","senderId":""")
            append(jsonString(routing.senderId))
            append(""","recipientId":""")
            append(jsonString(routing.recipientId))
            append(""","sentAt":${routing.sentAt},"sequence":${routing.sequence}}""")
        }
        return json.toByteArray(StandardCharsets.UTF_8)
    }

    private fun hmac(key: ByteArray, value: ByteArray): ByteArray {
        val mac = Mac.getInstance("HmacSHA256")
        mac.init(SecretKeySpec(key, "HmacSHA256"))
        return mac.doFinal(value)
    }

    private fun jsonString(value: String): String = buildString {
        for (character in value) {
            when (character) {
                '"' -> append("\\\"")
                '\\' -> append("\\\\")
                '\b' -> append("\\b")
                '\u000C' -> append("\\f")
                '\n' -> append("\\n")
                '\r' -> append("\\r")
                '\t' -> append("\\t")
                else -> {
                    if (character.code < 0x20) {
                        append("\\u%04x".format(character.code))
                    } else {
                        append(character)
                    }
                }
            }
        }
    }
}

class RelaySequenceWindow(initial: Map<String, Long> = emptyMap()) {
    private val highest = initial.toMutableMap()

    fun accept(senderId: String, sequence: Long) {
        if (sequence <= (highest[senderId] ?: 0)) {
            throw SecurityException("Relay replay rejected")
        }
        highest[senderId] = sequence
    }

    fun snapshot(): Map<String, Long> = highest.toMap()
}

File: mac/Sources/RelayCore/RelayCloudCrypto.swift
import CryptoKit
import Foundation

public enum RelayCloudCryptoError: Error, Equatable, Sendable {
    case invalidEnvelope
    case authenticationFailed
    case replay
}

public struct RelayTunnelRouting: Codable, Equatable, Sendable {
    public var version = 1
    public var messageID: String
    public var accountID: String
    public var hostID: String
    public var senderID: String
    public var recipientID: String
    public var sentAt: Int64
    public var sequence: Int64

    public init(
        messageID: String,
        accountID: String,
        hostID: String,
        senderID: String,
        recipientID: String,
        sentAt: Int64,
        sequence: Int64
    ) {
        self.messageID = messageID
        self.accountID = accountID
        self.hostID = hostID
        self.senderID = senderID
        self.recipientID = recipientID
        self.sentAt = sentAt
        self.sequence = sequence
    }

    enum CodingKeys: String, CodingKey {
        case version
        case messageID = "messageId"
        case accountID = "accountId"
        case hostID = "hostId"
        case senderID = "senderId"
        case recipientID = "recipientId"
        case sentAt
        case sequence
    }
}

public struct RelayTunnelEnvelope: Codable, Equatable, Sendable {
    public var version: Int
    public var messageID: String
    public var accountID: String
    public var hostID: String
    public var senderID: String
    public var recipientID: String
    public var sentAt: Int64
    public var sequence: Int64
    public var nonce: String
    public var ciphertext: String

    public var routing: RelayTunnelRouting {
        RelayTunnelRouting(
            messageID: messageID,
            accountID: accountID,
            hostID: hostID,
            senderID: senderID,
            recipientID: recipientID,
            sentAt: sentAt,
            sequence: sequence
        )
    }

    enum CodingKeys: String, CodingKey {
        case version
        case messageID = "messageId"
        case accountID = "accountId"
        case hostID = "hostId"
        case senderID = "senderId"
        case recipientID = "recipientId"
        case sentAt
        case sequence
        case nonce
        case ciphertext
    }
}

public enum RelayCloudCrypto {
    public static func deriveRootKey(
        privateKey: P256.KeyAgreement.PrivateKey,
        peerPublicKey: P256.KeyAgreement.PublicKey,
        pairingSessionNonce: Data
    ) throws -> SymmetricKey {
        guard pairingSessionNonce.count >= 16 else {
            throw RelayCloudCryptoError.invalidEnvelope
        }
        let secret = try privateKey.sharedSecretFromKeyAgreement(
            with: peerPublicKey
        )
        return secret.hkdfDerivedSymmetricKey(
            using: SHA256.self,
            salt: pairingSessionNonce,
            sharedInfo: Data("relay-e2ee-v1".utf8),
            outputByteCount: 32
        )
    }

    public static func encrypt(
        _ plaintext: Data,
        routing: RelayTunnelRouting,
        rootKey: SymmetricKey,
        nonce: Data
    ) throws -> RelayTunnelEnvelope {
        guard nonce.count == 12 else {
            throw RelayCloudCryptoError.invalidEnvelope
        }
        let sealed = try AES.GCM.seal(
            plaintext,
            using: rootKey,
            nonce: try AES.GCM.Nonce(data: nonce),
            authenticating: canonicalAAD(routing)
        )
        return RelayTunnelEnvelope(
            version: routing.version,
            messageID: routing.messageID,
            accountID: routing.accountID,
            hostID: routing.hostID,
            senderID: routing.senderID,
            recipientID: routing.recipientID,
            sentAt: routing.sentAt,
            sequence: routing.sequence,
            nonce: base64URL(nonce),
            ciphertext: base64URL(sealed.ciphertext + sealed.tag)
        )
    }

    public static func decrypt(
        _ envelope: RelayTunnelEnvelope,
        rootKey: SymmetricKey
    ) throws -> Data {
        do {
            guard
                envelope.version == 1,
                envelope.sequence > 0,
                let nonce = Data(base64URL: envelope.nonce),
                let combined = Data(base64URL: envelope.ciphertext),
                combined.count >= 16
            else {
                throw RelayCloudCryptoError.invalidEnvelope
            }
            let ciphertext = combined.dropLast(16)
            let tag = combined.suffix(16)
            let box = try AES.GCM.SealedBox(
                nonce: AES.GCM.Nonce(data: nonce),
                ciphertext: ciphertext,
                tag: tag
            )
            return try AES.GCM.open(
                box,
                using: rootKey,
                authenticating: canonicalAAD(envelope.routing)
            )
        } catch {
            throw RelayCloudCryptoError.authenticationFailed
        }
    }

    public static func canonicalAAD(_ routing: RelayTunnelRouting) -> Data {
        let json = """
        {"version":\(routing.version),"messageId":\(jsonString(routing.messageID)),"accountId":\(jsonString(routing.accountID)),"hostId":\(jsonString(routing.hostID)),"senderId":\(jsonString(routing.senderID)),"recipientId":\(jsonString(routing.recipientID)),"sentAt":\(routing.sentAt),"sequence":\(routing.sequence)}
        """
        return Data(json.utf8)
    }

    private static func jsonString(_ value: String) -> String {
        let data = try! JSONEncoder().encode(value)
        return String(decoding: data, as: UTF8.self)
    }

    private static func base64URL(_ data: Data) -> String {
        data.base64EncodedString()
            .replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_")
            .replacingOccurrences(of: "=", with: "")
    }
}

public struct RelayCloudReplayWindow: Equatable, Sendable {
    public private(set) var highestSequences: [String: Int64]

    public init(highestSequences: [String: Int64] = [:]) {
        self.highestSequences = highestSequences
    }

    public mutating func accept(senderID: String, sequence: Int64) throws {
        guard sequence > (highestSequences[senderID] ?? 0) else {
            throw RelayCloudCryptoError.replay
        }
        highestSequences[senderID] = sequence
    }
}

private extension Data {
    init?(base64URL: String) {
        var value = base64URL
            .replacingOccurrences(of: "-", with: "+")
            .replacingOccurrences(of: "_", with: "/")
        value += String(repeating: "=", count: (4 - value.count % 4) % 4)
        self.init(base64Encoded: value)
    }
}
